File: src/tachikoma/projects/git.py
"""Shared async git helpers for submodule operations.

Provides subprocess wrappers for all git operations needed by the projects
package: hooks, processor, tools, and context provider.
"""

import asyncio
import shutil
from pathlib import Path

from loguru import logger

_log = logger.bind(component="projects")


async def _run_git(*args: str, cwd: Path) -> None:
    """Run a git command and raise on failure.

    Args:
        *args: Git command arguments (e.g., "checkout", "main").
        cwd: Working directory for the command.

    Raises:
        RuntimeError: If the command returns non-zero exit code.
    """
    proc = await asyncio.create_subprocess_exec(
        "git",
        *args,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    _, stderr = await proc.communicate()

    if proc.returncode != 0:
        error_msg = stderr.decode().strip() or f"exit code {proc.returncode}"
        raise RuntimeError(f"git {' '.join(args)} failed: {error_msg}")


async def _run_git_capture(*args: str, cwd: Path) -> tuple[int, str]:
    """Run a git command and return exit code + stdout.

    Args:
        *args: Git command arguments.
        cwd: Working directory for the command.

    Returns:
        Tuple of (returncode, stdout as decoded string).
    """
    proc = await asyncio.create_subprocess_exec(
        "git",
        *args,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    stdout, _ = await proc.communicate()
    return proc.returncode or 0, stdout.decode().strip()


async def list_submodules(workspace: Path) -> list[str]:
    """List all submodule paths via `git submodule status`.

    Args:
        workspace: The workspace root directory.

    Returns:
        List of submodule paths (e.g., ["projects/my-app"]).
    """
    rc, output = await _run_git_capture("submodule", "status", "--recursive", cwd=workspace)

    if rc != 0:
        return []

    # Each line is like " abc1234 projects/my-app (heads/main)"
    # First character is a status indicator: space, +, -, or U
    paths: list[str] = []
    for line in output.splitlines():
        parts = line.strip().split()
        if len(parts) >= 2:
            paths.append(parts[1])

    return paths


async def init_submodule(workspace: Path, path: str) -> None:
    """Initialize a submodule via `git submodule update --init`.

    Args:
        workspace: The workspace root directory.
        path: The submodule path (e.g., "projects/my-app").

    Raises:
        RuntimeError: If the command fails.
    """
    await _run_git("submodule", "update", "--init", path, cwd=workspace)


async def resolve_default_branch(submodule_path: Path) -> str:
    """Resolve the default branch from the remote's HEAD reference.

    Tries `git symbolic-ref refs/remotes/origin/HEAD` first (local read).
    Falls back to `git remote show origin` (network call) if needed.
    Final fallback: return "main".

    Args:
        submodule_path: The submodule directory.

    Returns:
        The default branch name (e.g., "main", "master").
    """
    rc, output = await _run_git_capture(
        "symbolic-ref",
        "refs/remotes/origin/HEAD",
        cwd=submodule_path,
    )

    if rc == 0 and output.startswith("refs/remotes/origin/"):
        return output[len("refs/remotes/origin/") :]

    # Fallback: network call via remote show
    rc, output = await _run_git_capture("remote", "show", "origin", cwd=submodule_path)

    if rc == 0:
        for line in output.splitlines():
            if "HEAD branch:" in line:
                return line.split(":")[-1].strip()

    _log.warning(
        "Could not resolve default branch, using 'main': path={path}",
        path=str(submodule_path),
    )
    return "main"


async def checkout_branch(submodule_path: Path, branch: str) -> None:
    """Checkout a branch in the submodule.

    Args:
        submodule_path: The submodule directory.
        branch: The branch name to checkout.

    Raises:
        RuntimeError: If the command fails.
    """
    await _run_git("checkout", branch, cwd=submodule_path)


async def fetch(submodule_path: Path) -> None:
    """Fetch updates from the remote.

    Args:
        submodule_path: The submodule directory.

    Raises:
        RuntimeError: If the command fails.
    """
    await _run_git("fetch", cwd=submodule_path)


async def pull(submodule_path: Path) -> None:
    """Pull updates from the remote.

    Args:
        submodule_path: The submodule directory.

    Raises:
        RuntimeError: If the command fails.
    """
    await _run_git("pull", cwd=submodule_path)


async def has_uncommitted_changes_detail(submodule_path: Path) -> str | None:
    """Get detailed status of uncommitted changes.

    Args:
        submodule_path: The submodule directory.

    Returns:
        The git status --porcelain output, or None if clean.
    """
    _, output = await _run_git_capture("status", "--porcelain", cwd=submodule_path)
    return output if output else None


async def is_dirty(submodule_path: Path) -> bool:
    """Check if the submodule has uncommitted changes.

    Args:
        submodule_path: The submodule directory.

    Returns:
        True if there are uncommitted changes, False otherwise.
    """
    return await has_uncommitted_changes_detail(submodule_path) is not None


async def push(submodule_path: Path) -> None:
    """Push commits to the remote.

    Args:
        submodule_path: The submodule directory.

    Raises:
        RuntimeError: If the command fails.
    """
    await _run_git("push", cwd=submodule_path)


async def add_submodule(workspace: Path, name: str, url: str) -> None:
    """Add a new git submodule.

    Args:
        workspace: The workspace root directory.
        name: The project name (subdirectory under projects/).
        url: The git remote URL.

    Raises:
        RuntimeError: If the command fails.
    """
    await _run_git("submodule", "add", url, f"projects/{name}", cwd=workspace)


async def remove_submodule(workspace: Path, name: str) -> None:
    """Remove a git submodule completely.

    Runs: git submodule deinit -f projects/<name>
          git rm -f projects/<name>
          rm -rf .git/modules/projects/<name>

    Args:
        workspace: The workspace root directory.
        name: The project name (subdirectory under projects/).

    Raises:
        RuntimeError: If any command fails.
    """
    path = f"projects/{name}"

    await _run_git("submodule", "deinit", "-f", path, cwd=workspace)
    await _run_git("rm", "-f", path, cwd=workspace)

    modules_path = workspace / ".git" / "modules" / "projects" / name
    if modules_path.exists():
        shutil.rmtree(modules_path)


async def current_branch(submodule_path: Path) -> str | None:
    """Get the current branch name, or None if detached HEAD.

    Args:
        submodule_path: The submodule directory.

    Returns:
        The branch name, or None if in detached HEAD state.
    """
    rc, output = await _run_git_capture("symbolic-ref", "--short", "HEAD", cwd=submodule_path)

    if rc != 0:
        return None

    return output or None


async def current_commit_short(submodule_path: Path) -> str:
    """Get the short commit hash of HEAD.

    Args:
        submodule_path: The submodule directory.

    Returns:
        The short commit hash (e.g., "abc1234").
    """
    rc, output = await _run_git_capture("rev-parse", "--short", "HEAD", cwd=submodule_path)

    if rc != 0:
        return "unknown"

    return output or "unknown"

File: src/extensions/projects/git.ts
import { rm } from "node:fs/promises";
import { join } from "node:path";

import { runGit, runGitCapture } from "../../git/git.ts";
import { resolveRemoteDefaultBranch } from "../../git/remote.ts";
import { DIVERGENCE_STATUS, detectDivergence } from "../../git/sync.ts";

// `listSubmodules` is a core git primitive (it just parses `git submodule status`),
// so it lives in `src/git/git.ts` and is re-exported here for the project tools that
// already import it from this module.
export { listSubmodules } from "../../git/git.ts";

export const initSubmodule = async (workspaceRoot: string, path: string): Promise<void> => {
  await runGit(workspaceRoot, ["submodule", "update", "--init", path]);
};

/**
 * Resolve the default branch through the shared core resolver (local symbolic ref, then
 * `ls-remote --symref`), defaulting to "main" when both fail — project tooling is
 * best-effort and must keep working against an unreachable or empty remote, unlike
 * skill-evolution's strict must-abort resolution.
 */
export const resolveDefaultBranch = async (repoPath: string): Promise<string> => {
  try {
    return await resolveRemoteDefaultBranch(repoPath);
  } catch {
    return "main";
  }
};

export const checkoutBranch = async (repoPath: string, branch: string): Promise<void> => {
  await runGit(repoPath, ["checkout", branch]);
};

/** Porcelain status output, or null when the working tree is clean. */
export const uncommittedChangesDetail = async (repoPath: string): Promise<string | null> => {
  const { stdout } = await runGitCapture(repoPath, ["status", "--porcelain"]);
  return stdout === "" ? null : stdout;
};

export const isDirty = async (repoPath: string): Promise<boolean> =>
  (await uncommittedChangesDetail(repoPath)) != null;

export const addSubmodule = async (
  workspaceRoot: string,
  name: string,
  url: string,
): Promise<void> => {
  await runGit(workspaceRoot, ["submodule", "add", url, `projects/${name}`]);
};

/** Remove a submodule completely: deinit, git rm, and drop the .git/modules clone. */
export const removeSubmodule = async (workspaceRoot: string, name: string): Promise<void> => {
  const path = `projects/${name}`;

  await runGit(workspaceRoot, ["submodule", "deinit", "-f", path]);
  await runGit(workspaceRoot, ["rm", "-f", path]);
  await rm(join(workspaceRoot, ".git", "modules", "projects", name), {
    recursive: true,
    force: true,
  });
};

/** Current branch name, or null when HEAD is detached. */
export const currentBranch = async (repoPath: string): Promise<string | null> => {
  const result = await runGitCapture(repoPath, ["symbolic-ref", "--short", "HEAD"]);

  if (result.code !== 0 || result.stdout === "") return null;

  return result.stdout;
};

/**
 * Whether HEAD has commits not on the configured `origin` branch — a cheap,
 * fetch-free check against the last-known remote-tracking ref. Returns false for
 * detached HEAD or when no `origin/<branch>` ref exists (no remote / never
 * fetched), so callers only push submodules that genuinely look ahead.
 */
export const isAhead = async (repoPath: string): Promise<boolean> => {
  const branch = await currentBranch(repoPath);

  if (branch == null) return false;

  return (await detectDivergence(repoPath, "origin", branch)) === DIVERGENCE_STATUS.ahead;
};

export const currentCommitShort = async (repoPath: string): Promise<string> => {
  const result = await runGitCapture(repoPath, ["rev-parse", "--short", "HEAD"]);

  if (result.code !== 0 || result.stdout === "") return "unknown";

  return result.stdout;
};

export interface ProjectState {
  name: string;
  /** Branch name, or null when HEAD is detached. */
  branch: string | null;
  commit: string;
  dirtyFiles: number;
}

/** Gather display state (branch/commit + dirty file count) for one submodule path. */
export const projectState = async (workspaceRoot: string, path: string): Promise<ProjectState> => {
  const repoPath = join(workspaceRoot, path);
  const detail = await uncommittedChangesDetail(repoPath);

  return {
    name: path.split("/").at(-1) ?? path,
    branch: await currentBranch(repoPath),
    commit: await currentCommitShort(repoPath),
    dirtyFiles: detail == null ? 0 : detail.split("\n").length,
  };
};

export const describeProjectState = (state: ProjectState): string => {
  const location = state.branch ?? `${state.commit} (detached)`;
  const dirty =
    state.dirtyFiles > 0
      ? ` — ${state.dirtyFiles} uncommitted change${state.dirtyFiles === 1 ? "" : "s"}`
      : "";

  return `- ${state.name}: ${location}${dirty}`;
};
